compiler/daml-lf-tools/src/DA/Daml/LF/TypeChecker/NameCollision.hs

```diff
@@ -0,0 +1,162 @@
+-- Copyright (c) 2019 The DAML Authors. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+module DA.Daml.LF.TypeChecker.NameCollision
+    ( checkModule
+    ) where
+
+import DA.Daml.LF.Ast
+import DA.Daml.LF.TypeChecker.Env
+import DA.Daml.LF.TypeChecker.Error
+import Data.Maybe
+import Control.Monad.Extra
+import qualified Data.Map.Strict as M
+import qualified Data.Text as T
+import qualified Control.Monad.Trans.State as S
+
+-- | The various names we wish to track within a package.
+-- This type separates all the different kinds of names
+-- out nicely, and preserves case sensitivity, so the
+-- names are easy to display in error messages. To get
+-- the corresponding case insensitive fully resolved name,
+-- see 'FRName'.
+data Name
+    = NModule ModuleName
+    | NRecordType ModuleName TypeConName
+    | NVariantType ModuleName TypeConName
+    | NEnumType ModuleName TypeConName
+    | NVariantCon ModuleName TypeConName VariantConName
+    | NEnumCon ModuleName TypeConName VariantConName
+    | NField ModuleName TypeConName FieldName
+    | NChoice ModuleName TypeConName ChoiceName
+
+-- | Display a name in a super unambiguous way.
+displayName :: Name -> T.Text
+displayName = \case
+    NModule (ModuleName m) ->
+        T.concat ["module ", dot m]
+    NRecordType (ModuleName m) (TypeConName t) ->
+        T.concat ["record ", dot m, ":", dot t]
+    NVariantType (ModuleName m) (TypeConName t) ->
+        T.concat ["variant ", dot m, ":", dot t]
+    NEnumType (ModuleName m) (TypeConName t) ->
+        T.concat ["enum ", dot m, ":", dot t]
+    NVariantCon (ModuleName m) (TypeConName t) (VariantConName v) ->
+        T.concat ["variant constructor ", dot m, ":", dot t, ".", v]
+    NEnumCon (ModuleName m) (TypeConName t) (VariantConName v) ->
+        T.concat ["enum constructor ", dot m, ":", dot t, ".", v]
+    NField (ModuleName m) (TypeConName t) (FieldName f) ->
+        T.concat ["field ", dot m, ":", dot t, ".", f]
+    NChoice (ModuleName m) (TypeConName t) (ChoiceName c) ->
+        T.concat ["choice ", dot m, ":", dot t, ".", c]
+  where
+    dot = T.intercalate "."
+
+-- | Asks whether a name collision is permitted. According to the
+-- LF Spec, a name collision is only permitted when it occurs
+-- between a record type and a variant constructor defined in
```

Comment:
That seems like a weird rule, I would have expected the opposite if anything, i.e., collisions are only allowed if they are in different modules …

Reply:
I think it's got something to do with the desugaring of named fields in variant types.

Reply:
Let consider the following daml:
```
module Mod where
data T @a : A {x: a, y: a} | B {x: a, y: a} = ..
```

is compiled in LF
```
mod Mod {
  
record T.A @a = {x: a, y: a} 
record T.B @a = {x: a, y: a} 

variant T @a := A (Mod:T.A a) | B (Mod:T.B b)
```
 
In this case you have a collision between

- the record constructor `A.C` (noted `Mod:A.C` in LF and translated in java/scala as `Mod.A.C`) and 
- the constructor `A` of the variant `T` (noted `Mod:A:C` in LF and translate in java/scala as `Mod.A.C`).

This is OK to have such "collision" because the record constructor and the variant constructor will be merged back in the same object in java/scala.

Reply:
Thanks for the explanation @remyhaemmerle-da!